sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,652 @@ These follow the AAELF64_ base definition from Arm.
+   static link time value.
+
+ * The IFUNC resolver function for a given function must be defined in
+   the same translation unit as the implementations of the function.
+
+ * IFUNC resolver functions must be idempotent. There can be
+   relocations in both ``.rela.dyn`` and ``.rela.plt`` to the same
+   IFUNC resolver function.
+
+IFUNC requirements for static linkers
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Relocations to pre-emptable symbols of type ``STT_GNU_IFUNC`` are
+handled in the same way as symbols of type ``STT_FUNC``. The symbol
+type ``STT_GNU_IFUNC`` is propagated into the dynamic symbol table.
+
+Relocations to non-pre-empatable symbols of type ``STT_GNU_IFUNC`` are
+resolved to a PLT entry that loads the value of:
+
+ * A ``.got.plt`` entry for a non-branch relocation such as
+   ``R_AARCH64_ABS64``. The entry has an ``R_AARCH64_IRELATIVE``
+   dynamic relocation in ``.rela.dyn``.
+
+ * A ``.got.plt`` entry for a branch relocation such as
+   ``R_AARCH64_CALL``. A ``R_AARCH64_IRELATIVE`` dynamic relocation is
+   added to ``.rela.dyn`` or ``rela.plt``.
+
+Due to the ordering requirements on IFUNC resolvers the PLT entry and
+associated ``.got.plt`` entry are often implemented in a separate
+``.iplt`` and ``.iplt.got`` sections so that they are placed after the
+``.plt`` and ``.got.plt`` sections respectively.
+
+Like ``R_AARCH64_RELATIVE`` the ``R_AARCH64_IRELATIVE`` relocation
+does not require a symbol. For ``RELA`` type relocations the addend of
+the ``R_AARCH64_IRELATIVE`` relocation contains the address of the
+IFUNC resolver function.
+
+To make address equivalence of functions with IFUNC resolvers work, if
+the address of a non-preemptable ``STT_GNU_IFUNC`` symbol is taken in
+a non-position-independent executable. The static linker must use the
+address of the corresponding PLT entry for the address of the
+function. If the symbol is exported to the dynamic symbol table the
+``st_value`` of the symbol must be set to the address of the PLT entry
+and the dynamic symbol table type must be set to ``STT_FUNC``.
+
+Relocations of type ``R_AARCH64_IRELATIVE`` relocation must be sorted
+after all other relocation types. This means that for a given
+executable or shared-library the following ordering can be relied on:
+
+ * All non ``R_AARCH64_IRELATIVE`` relocations in ``.rela.dyn`` will be resolved
+   before any ``R_AARCH64_IRELATIVE`` relocations.
+
+ * ALL IFUNC resolvers with ``R_AARCH64_IRELATIVE`` relocations in
+   ``.rela.dyn`` will be run before the ``.rela.plt`` relocations are
+   resolved.
+
+ * All non ``R_AARCH64_IRELATIVE`` relocations in ``.rela.plt`` will be resolved
+   before any ``R_AARCH64_IRELATIVE`` relocations.
+
+ * All IFUNC resolvers with ``R_AARCH64_IRELATIVE`` relocations in
+   ``.rela.plt`` will be run last.
+
+When static linking, the dynamic relocation section containing
+``R_AARCH64_IRELATIVE`` relocations are output as if dynamic
+linking. The static linker must define the symbols
```

Comment:
`__rela_iplt_start` is only defined for `-no-pie`, not for `-pie`.

See https://github.com/llvm/llvm-project/commit/f8cb78e99aae9aa3f89f7bfe667db2c5b767f21f (`[ELF] Don't define __rela_iplt_start for -pie/-shared`)

Reply:
Thanks for the link. I'll update to account for that.